Ignore least preferred servers if all servers are least preferred

tiwillia opened the pull request:
Bug 1228373
Bug link https://bugzilla.redhat.com/show_bug.cgi?id=1228373

If an application is already scaled out so that it has a gear on all available nodes, all nodes will be considered 'least_preferred_servers'. This will cause only one server to be returned from rpc_find_all_available, giving select_best_fit_node only one server to choose. This caused sparatic and uneven gear placement from highly scaled applications.

This fix will ignore the least_preferred_servers list if all available servers are included in the list.

tiwillia committed:
Ignore least preferred servers if all servers are least preferred

Bug 1228373
Bug link https://bugzilla.redhat.com/show_bug.cgi?id=1228373

If an application is already scaled out so that it has a gear on all available nodes, all nodes will be considered 'least_preferred_servers'. This will cause only one server to be returned from rpc_find_all_available, giving select_best_fit_node only one server to choose. This caused sparatic and uneven gear placement from highly scaled applications.

This fix will ignore the least_preferred_servers list if all available servers are included in the list.

tiwillia commented:
[test]

danmcp commented on plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb:
```diff
@@ -3222,14 +3222,25 @@ def self.rpc_find_all_available(opts=nil)
+            if least_preferred_servers.present?
+              ignore_least_preferred_servers = false
+              server_identities = server_infos.map {|server_info| server_info.name}.sort
+              if least_preferred_servers.compact.sort == server_identities
```
You can't be sure that all servers are returned in server_infos for multiple reasons including:
- They won't be in the list if they are full
- They won't be in the list if the missed the mco discovery timeout.

tiwillia commented on plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb:
```diff
@@ -3222,14 +3222,25 @@ def self.rpc_find_all_available(opts=nil)
+            if least_preferred_servers.present?
+              ignore_least_preferred_servers = false
+              server_identities = server_infos.map {|server_info| server_info.name}
+              if server_identities.all? {|identity| least_preferred_servers.include?(identity)}
```
@danmcp Could you review again?

danmcp commented on plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb:
```diff
@@ -3222,14 +3222,25 @@ def self.rpc_find_all_available(opts=nil)
 
+            # Ignore least_preferred_servers if all servers are considered least preferred
+            if least_preferred_servers.present?
+              ignore_least_preferred_servers = false
```
Seems better if this line is moved up 1.

danmcp commented on plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb:
```diff
@@ -3251,7 +3262,12 @@ def self.rpc_find_all_available(opts=nil)
-        server_infos.delete_if { |server_info| (server_infos.length > 1) && least_preferred_servers.include?(server_info.name) } if least_preferred_servers.present?
+        # Ignore least preferred servers if all servers are considered least preferred
+        if least_preferred_servers.present?
+          unless ignore_least_preferred_servers
```
Do this in one line as 

if least_preferred_servers.present? && !ignore_least_preferred_servers

danmcp commented on plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb:
```diff
@@ -3251,7 +3262,12 @@ def self.rpc_find_all_available(opts=nil)
+        # Ignore least preferred servers if all servers are considered least preferred
+        if least_preferred_servers.present?
+          unless ignore_least_preferred_servers
+            server_infos.delete_if { |server_info| (server_infos.length > 1) && least_preferred_servers.include?(server_info.name) }
```
So if this works the > 1 check will never fail right?

danmcp commented on plugins/msg-broker/mcollective/lib/openshift/mcollective_application_container_proxy.rb:
```diff
@@ -3222,14 +3222,25 @@ def self.rpc_find_all_available(opts=nil)
+                least_preferred_zone_ids << server.zone_id if server.zone_id
+              end
+              least_preferred_zone_ids = least_preferred_zone_ids.uniq
+            end
```
Should this also include the block below?

tiwillia commented:
Moved where `ignore_least_preferred_servers` was set. If a region was not defined, it would never be set, meaning that all servers were deleted from `server_infos` at the end of the method. Tested again with current code:

District configuration:
```
{"_id"=>"557afe8d84c8718aa0000001",
 "active_servers_size"=>4,
 "available_capacity"=>5900,
 "available_uids"=>"<5900 uids hidden>",
 "created_at"=>2015-06-12 15:45:17 UTC,
 "gear_size"=>"small",
 "max_capacity"=>6000,
 "max_uid"=>6999,
 "name"=>"small-district",
 "platform"=>"linux",
 "servers"=>
  [{"_id"=>"557afeb484c87111af000001",
    "active"=>true,
    "name"=>"ip-10-99-165-109",
    "region_id"=>"557aff0184c8713199000001",
    "region_name"=>"RDU",
    "unresponsive"=>false,
    "zone_id"=>"557aff3b84c871dbb3000001",
    "zone_name"=>"south"},
   {"_id"=>"557afebf84c871a72b000001",
    "active"=>true,
    "name"=>"ip-10-61-190-123",
    "region_id"=>"557aff0184c8713199000001",
    "region_name"=>"RDU",
    "unresponsive"=>false,
    "zone_id"=>"557aff2d84c871fd9c000001",
    "zone_name"=>"north"},
   {"_id"=>"557afec884c871dbb9000001",
    "active"=>true,
    "name"=>"ip-10-71-180-79",
    "region_id"=>"557aff0184c8713199000001",
    "region_name"=>"RDU",
    "unresponsive"=>false,
    "zone_id"=>"557aff2d84c871fd9c000001",
    "zone_name"=>"north"},
   {"_id"=>"557b017584c8718d1f000001",
    "active"=>true,
    "name"=>"ip-10-16-180-140",
    "region_id"=>"557aff0184c8713199000001",
    "region_name"=>"RDU",
    "unresponsive"=>false,
    "zone_id"=>"557aff3b84c871dbb3000001",
    "zone_name"=>"south"}],
 "updated_at"=>2015-06-12 15:45:17 UTC,
 "uuid"=>"066ef74a111a11e5a5360242ac110002"}
```
Created 1 ruby-1.9 application and scaled to a minimum of 100. Gear counts across nodes:

Hostname | zone | gears
--------------|--------|----------
ip-10-71-180-79  | north | 25 
ip-10-61-190-123 | north | 39 
ip-10-99-165-109 | south | 17
ip-10-16-180-140 | south | 19

Removed regions/zones, tested again with one application scaled to 100 gears:


Hostname | gears
--------------|--------
ip-10-71-180-79   | 30 
ip-10-61-190-123 | 23 
ip-10-99-165-109 | 14
ip-10-16-180-140 | 33

Gear placement is not nearly as erratic.

tiwillia commented:
[test] evaluation results appear to be unrelated

tiwillia commented:
[test]

danmcp commented:
[merge]
